File: 1.singly_linked_lists/singly_linked_list.py

class Node:
    def __init__(self, value):
        self.value = value
        self.next_node = None

class Linkedlist:
    def __init__(self):
        self.head = None

#   -------------------------------------------------------------------------------------------
    def is_list_empty(self):
        return self.head == None

#   -------------------------------------------------------------------------------------------

    def add_node_to_front(self, value):
        new_node = Node(value)
        new_node.next_node = self.head
        self.head = new_node

#   -------------------------------------------------------------------------------------------

    def remove_node_from_front(self):
        if(self.is_list_empty()):
            print("List is empty!!!")
        else:
            temp = self.head
            self.head = temp.next_node
            del temp

#   -------------------------------------------------------------------------------------------

    def add_node_to_back(self, value):
        if(self.is_list_empty()):
            new_node = Node(value)
            self.head = new_node
            return
    
        temp = self.head
        while(temp.next_node != None):
            temp= temp.next_node
        new_node = Node(value)
        temp.next_node = new_node

#   -------------------------------------------------------------------------------------------

    def remove_node_from_back(self):
        if(self.is_list_empty()):
            print("List is empty!!!")
            
        else:
            if(self.head.next_node == None):
                del self.head
                self.head = None
            else:
                temp = self.head
                while((temp.next_node).next_node != None):
                    temp = temp.next_node
                del temp.next_node
                temp.next_node = None

#   -------------------------------------------------------------------------------------------

    def print_list(self):

        if(self.is_list_empty()):
            print("List is empty!!!")
        else:
            temp = self.head
            while(temp.next_node != None):
                print(f"{temp.value} -> ", end="")
                temp = temp.next_node
            print(f"{temp.value}")                

#   -------------------------------------------------------------------------------------------

    def add_item_by_order(self, value):
        new_node = Node(value)

        if(self.is_list_empty()):
            self.head = new_node
        
        else:
            temp = self.head
            
            if(value < temp.value):
                new_node.next_node = temp
                self.head = new_node
                return
        
            while(temp.next_node != None and value >= (temp.next_node).value):
                temp = temp.next_node
            
            new_node.next_node = temp.next_node
            temp.next_node = new_node

#   -------------------------------------------------------------------------------------------

    def remove_specific_node(self, value):
        if(self.is_list_empty()):
            print("List is empty!!!")
        
        else:
            if(self.head.value == value):
                temp = self.head
                self.head = self.head.next_node
                del temp
                return

            temp = self.head
            temp2 = self.head.next_node

            while(temp2 != None):
                if(temp2.value == value):
                    temp.next_node = temp2.next_node
                    break 
                temp = temp.next_node
                temp2 = temp2.next_node
            
            if(temp2 == None):
                print(f"There is not item with a value of {value}")
                return

            del temp2
            
#   -------------------------------------------------------------------------------------------

    def merge_two_list(self, list2):
        merged_list = Linkedlist()
        list1_tmp_head = self.head
        list2_tmp_head = list2.head

        while(list1_tmp_head != None or list2_tmp_head != None):
            if(list1_tmp_head != None and list2_tmp_head != None):
                while(list1_tmp_head != None and list2_tmp_head != None):
                    if(list1_tmp_head.value < list2_tmp_head.value):
                        merged_list.add_node_to_back(list1_tmp_head.value)
                        list1_tmp_head = list1_tmp_head.next_node
                    else:
                        merged_list.add_node_to_back(list2_tmp_head.value)
                        list2_tmp_head = list2_tmp_head.next_node

            if(list1_tmp_head == None):
                while(list2_tmp_head != None):
                    merged_list.add_node_to_back(list2_tmp_head.value)
                    list2_tmp_head = list2_tmp_head.next_node

            if(list2_tmp_head == None):
                while(list1_tmp_head != None):
                    merged_list.add_node_to_back(list1_tmp_head.value)
                    list1_tmp_head = list1_tmp_head.next_node
                    
        return merged_list

    def merge_two_list_by_order(self, list2):
        merged_list = Linkedlist()
        list1_tmp_head = self.head
        list2_tmp_head = list2.head

        while(list1_tmp_head != None):
            merged_list.add_item_by_order(list1_tmp_head.value)
            list1_tmp_head = list1_tmp_head.next_node
                
        
        while(list2_tmp_head != None):
            merged_list.add_item_by_order(list2_tmp_head.value)
            list2_tmp_head = list2_tmp_head.next_node
        merged_list.print_list()
        return merged_list
        #     if(self.head == None):
        #         merged_list.add_node_to_front()

#   -------------------------------------------------------------------------------------------
#   -------------------------------------------------------------------------------------------

list1 = Linkedlist()
list2 = Linkedlist()

list1.add_node_to_front(6)
list1.add_node_to_back(7)
list1.add_node_to_front(3)
list1.add_node_to_front(1)
list1.add_node_to_front(57)
list1.add_node_to_front(723)
list1.add_node_to_front(74)
list1.add_node_to_back(3)
list1.add_node_to_back(0)
list1.add_node_to_back(5)
list1.add_node_to_front(5)
list1.add_node_to_back(65)
list1.add_node_to_front(1)
list1.add_node_to_back(57)
list1.add_node_to_back(65)




list2.add_node_to_front(0)
list2.add_node_to_front(5)
list2.add_node_to_front(65)



list1.print_list()
list2.print_list()
list3 = list1.merge_two_list(list2)
list4 = list1.merge_two_list_by_order(list2)
list3.print_list()
# list4.print_list()

File: 1.singly_linked_lists/singly_linked_list.c
#include <stdio.h>
#include <stdlib.h>


typedef struct Node{
    int value;
    struct Node *next;
}Node;



Node *head = NULL;

int is_list_empty(){
    if(head == NULL)
        return 1;
    
    else
        return 0;
}


void add_node_to_front(int value){
    Node *new_node = (Node *)malloc(sizeof(Node));
    new_node->value = value;
    new_node->next = head;
    head = new_node;

}


void remove_node_from_front(){
    if(is_list_empty()){
        printf("List Is Empty!!!\n");
        return;
    }
    Node *temp = head;
    head = head->next;
    free(temp);
}


void add_node_to_back(int value){
    Node *new_node = (Node *)malloc(sizeof(Node));
    new_node->value = value;
    new_node->next = NULL;

    if(is_list_empty()){
        head = new_node;
    }
    else{
        Node *temp = head;
        while(temp->next != NULL){
            temp = temp->next;
        }

        temp->next = new_node;
    }
}

void remove_node_from_back(){
    if(is_list_empty()){
        printf("List Is Empty!!!\n");
    }

    else if( head->next == NULL){
        free(head);
        head = NULL;
    }

    else{
        Node *temp = head;
        Node *temp2= head->next;
        while(temp2->next != NULL){
            temp = temp->next;
            temp2 = temp2->next;
            
        }
        free(temp2);
        temp->next = NULL;
    }
    
}


void add_item_by_order(int value){
    Node *new_node = (Node *)malloc(sizeof(Node));
    new_node->value = value;
    new_node->next = NULL;
    if(is_list_empty()){
        head = new_node;
    }
    else{
        Node *temp = head;
        if(temp->value > value){
            new_node->next = head;
            head = new_node;
            return;
        }

        while(temp->next != NULL && (temp->next)->value < value){
            temp = temp->next;
        }
        new_node->next = temp->next;
        temp->next = new_node;

    }
}


void remove_specific_node(int value){
    if(is_list_empty()){
        printf("List Is Empty!!!\n");
        return;
    }
    else{
        Node *temp = head;
        Node *temp2 = head->next;

        if(head->value == value){
            head = temp2;
            free(temp);
            return;
        }   
        while(temp2 != NULL){
            if(temp2->value == value){
                temp->next = temp2->next;
                free(temp2);
                return;
            }
            temp = temp->next;
            temp2 = temp2->next;
        }
        if(temp2 == NULL){
            printf("There Is No Such a Node With The Given Value.\n");
        }
    }

}


int find_length() {
   int length = 0;
   Node *current;
    for(current = head; current != NULL; current = current->next) {
      length++;
    }
   return length;
}

void print_list(){
    if(is_list_empty()){
       printf("List Is Empty!!!\n");
        return;
    }

    Node *temp = head;
    while(temp->next != NULL){
        printf("%d -> ", temp->value);
        temp = temp->next;
    }
    printf("%d\n", temp->value);
    
}

int main(int argc, char *argv[]){
    
    add_item_by_order(5);
    add_item_by_order(3);
    add_item_by_order(7);
    add_item_by_order(8);
    add_item_by_order(54);
    print_list();
    // remove_node_from_front();
    add_item_by_order(0);
    int b = find_length();
    print_list();
    // remove_node_from_back();
    // remove_node_from_back();
    // remove_node_from_back();
    // print_list();
    remove_specific_node(0);

    print_list();

    int a = find_length();
    printf("%d, %d ", a, b);
   
}